remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/RemoteQueryAccessImpl.java

```diff
@@ -0,0 +1,70 @@
+package org.infinispan.query.remote.impl;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.api.Experimental;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.infinispan.marshall.protostream.impl.SerializationContextRegistry;
+import org.infinispan.protostream.ImmutableSerializationContext;
+import org.infinispan.protostream.ProtobufUtil;
+import org.infinispan.query.core.impl.QueryResultImpl;
+import org.infinispan.query.dsl.QueryResult;
+import org.infinispan.tasks.query.RemoteQueryAccess;
+
+@Scope(Scopes.NAMED_CACHE)
+@Experimental
+public class RemoteQueryAccessImpl implements RemoteQueryAccess {
+
+   @Inject
+   RemoteQueryManager remoteQueryManager;
+
+   @Inject
+   SerializationContextRegistry serializationContextRegistry;
+
+   @Inject
+   AdvancedCache<?, ?> cache;
+
+   @Override
+   public QueryResult<?> executeQuery(String queryString, Map<String, Object> namedParametersMap, Integer offset,
+                                      Integer maxResults, Integer hitCountAccuracy, boolean isLocal) throws IOException {
+      BaseRemoteQueryManager.QueryResultWithProjection result = remoteQueryManager.localQuery(queryString, namedParametersMap, offset, maxResults,
+            hitCountAccuracy, cache, isLocal);
+
+      QueryResult<Object> queryResult = result.queryResult();
+      String[] projection = result.projection();
+      List<Object> results = extractResults(serializationContextRegistry.getUserCtx(), projection != null, queryResult.list());
+      return new QueryResultImpl<>(queryResult.count(), results);
+   }
+
+   private static List<Object> extractResults(ImmutableSerializationContext serializationContext, boolean projection,
+                                        List<Object> results) throws IOException {
+      List<Object> unwrappedResults = new ArrayList<>(results.size());
+      if (projection) {
+         for (Object result : results) {
+            Object[] objects = (Object[]) result;
+            for (int i=0; i<objects.length; i++) {
+               Object value = objects[i];
+               if (serializationContext != null && value instanceof byte[]) {
+                  objects[i] = ProtobufUtil.fromWrappedByteArray(serializationContext, (byte[]) value);
+               }
+            }
+            unwrappedResults.add(objects);
+         }
+      } else {
+         unwrappedResults = new ArrayList<>(results.size());
+         for (Object r : results) {
+            if (serializationContext != null && r instanceof byte[]) {
+               r = ProtobufUtil.fromWrappedByteArray(serializationContext, (byte[]) r);
+            }
```

Comment:
nitpick
this code can be in a private method to avoid duplication
```java
   private static Object unmarshall(ImmutableSerializationContext context, Object result) throws IOException {
      return result instanceof byte[] bytes ?
            ProtobufUtil.fromWrappedByteArray(context, bytes) :
            result;
   }
```